polkadot/runtime/common/src/integration_tests.rs

```diff
@@ -1725,3 +1731,183 @@ fn cant_bid_on_existing_lease_periods() {
+		// deposit for para registration
+		let expected_deposit_para_registration: Balance = ParaDeposit::get() +
+			DataDepositPerByte::get() * validation_code_storage_size as Balance +
+			DataDepositPerByte::get() * genesis_storage_size as Balance;
+
+		assert_ok!(Registrar::reserve(signed(1)));
+		assert_ok!(Registrar::register(
+			signed(1),
+			para_id,
+			test_genesis_head(genesis_storage_size),
+			validation_code.clone(),
+		));
+		conclude_pvf_checking::<Test>(&validation_code, VALIDATORS, START_SESSION_INDEX);
+
+		// Para 2000 should be onboarding
+		assert_eq!(Paras::lifecycle(para_id), Some(ParaLifecycle::Onboarding));
+
+		// Start a new auction in the future
+		let duration = 99u32;
+		let lease_period_index_start = 4u32;
+		assert_ok!(Auctions::new_auction(
+			RuntimeOrigin::root(),
+			duration,
+			lease_period_index_start
+		));
+
+		// 2 sessions later 2000 is an on-demand parachain
+		run_to_session(START_SESSION_INDEX + 2);
+		assert_eq!(Paras::lifecycle(para_id), Some(ParaLifecycle::Parathread));
+
+		// Open a crowdloan for Para 1 for slots 0-3
+		assert_ok!(Crowdloan::create(
+			signed(1),
+			para_id,
+			1_000_000,                    // Cap
+			lease_period_index_start + 0, // First Slot
+			lease_period_index_start + 3, // Last Slot
+			200,                          // Block End
+			None,
+		));
+		let fund = Crowdloan::funds(para_id).unwrap();
+		let crowdloan_account = Crowdloan::fund_account_id(fund.fund_index);
+
+		// Bunch of contributions
+		let mut total = 0;
+		for i in 10..20 {
+			Balances::make_free_balance_be(&account_id(i), 1_000_000_000);
+			assert_ok!(Crowdloan::contribute(signed(i), para_id, 900 - i, None));
+			total += 900 - i;
+		}
+		// Go to end of auction where everyone won their slots
+		run_to_block(200);
+
+		assert_eq!(
+			Balances::reserved_balance(&account_id(1)),
+			expected_deposit_para_registration + SubmissionDeposit::get()
+		);
+		// crowdloan fund is reserved
+		assert_eq!(Balances::reserved_balance(&crowdloan_account), total);
+		// Crowdloan is appropriately set
+		assert!(Crowdloan::funds(para_id).is_some());
+
+		// New leases will start on block 400
+		let lease_start_block = 400;
+		run_to_block(lease_start_block);
+
+		// Slots are won by Para 1
+		assert!(!Slots::lease(para_id).is_empty());
+
+		// 2 sessions later it is a parachain
+		run_to_block(lease_start_block + 20);
+		assert_eq!(Paras::lifecycle(para_id), Some(ParaLifecycle::Parachain));
+
+		// lease is active and not near end
+		assert_noop!(
+			Slots::early_lease_refund(signed(1), para_id),
+			SlotsError::<Test>::PreconditionNotMet
+		);
+
+		// Cant dissolve yet
+		assert_noop!(
+			Crowdloan::dissolve(signed(1), para_id),
+			CrowdloanError::<Test>::NotReadyToDissolve
+		);
+
+		let lease_end_block = 8 * LeasePeriod::get() + LeaseOffset::get();
+		// early refund can be applied from the 1st block in the last lease period.
+		let early_refund_start_block = lease_end_block - LeasePeriod::get() + 1;
```

Comment:
Maybe a check on the boundary would be good. ie. 

```rust
run_to_block(lease_end_block - LeasePeriod::get());
assert_noop!(Slots::early_lease_refund(signed(1), para_id));
```

Reply:
Good point, added!